### ISSUE TITLE
Asynchronous REST calls

### ISSUE THREAD

**Issue by ChristianGruen** [BaseXdb/basex#989]
See http://www.mail-archive.com/basex-talk%40mailman.uni-konstanz.de/msg04405.html

Solutions provided by other systems:

* [util:eval-async](http://exist-db.org/exist/apps/fundocs/view.html?uri=http://exist-db.org/xquery/util) (eXist-db)
* [asynchronous-jobs](http://www.28.io/documentation/latest/modules/parallelism/asynchronous-jobs) (28msec)
* [xdmp:spawn](http://docs.marklogic.com/xdmp:spawn) (MarkLogic)

**Comment by mgaerber** [comment]
+1 for  proc:fork

More complex situations would require something like a real job queue anyways. Some lighter solutions rely on Redis. See https://github.com/gresrun/jesque for something that could be integrated in basex on basis of java extensions.

**Comment by apb2006** [comment]
I don't see what proc:fork gives you over using the shell to run a command in the background with proc:execute?.

This may be feature creep but Asynchronous calls are a subset of scheduled calls - run once now. I wonder if something like https://github.com/anderswisch/cron-expression might form a component. I am looking at https://github.com/anderswisch/cron-expression/blob/master/src/test/java/cron/CronScheduleTest.java#L93
and thinking the runnable would execute some xquery string. That could be done directly or via the  client API.

**Comment by apb2006** [comment]
I have taken a stab at this (work in progress)...
https://github.com/Quodatum/async/releases/tag/v0.0.7

I know little about Java Multithreading :-) so this code might do serious damage! 
It is really above exploring what kind of API would be useful to me. Lots of cleanup and documentation is required. I am interested in the idea of "Futures" in XQuery.

**Comment by ChristianGruen** [comment]
Thanks for the contribution! Be sure I'll have a look at it. It may take some time, though (I'll be on vacation as of next week).

**Comment by james-jw** [comment]
I think the idea of having an API for asynchronous calls would be a dream and I personally do not think forking a process and forking a thread are equivalent nor would I consider this feature creep. With a new thread some level of context can be maintained while a forked process has to more or less start over id imagine.

Optimally a future with promise like structure would be best for XQuery 3.1 in my imagination giving its functional nature. So for example using the future pattern:

```xquery
let $req := <http:request method="GET" login="async" password="isAwesome" />
let $worker := function ($reqIn, $uri) {
    http:send-request($reqIn, $uri)
}
return
   for $uri in $uris 
   let $future := proc:fork($worker, ($req, $uri))
   return
      $future() 
```
Alternatively, the promise pattern should also be supported for chaining processes in a new fork. For example:

```xquery
(: Same $worker, $req as above etc... :)
let $extractListItems := function ($res as map(*)) { $res?list?* } as item()*
let $error := function ($result as item()*) {
     trace($result, 'Request failed!')
}
return
   for $uri in $uris 
   let $retrieve:= 
                proc:fork($worker, ($req, $uri), map { 
                       'then': parse-json(?), 
                       'fail': $error 
                 })
    let $extract := $retrieve(map { 'then': $extractListItems  })
    return $extract()
```
As you can see, passing in a map to a forked function adds callbacks instead of retrieving the result.

The supported callbacks would be:
* done
* then
* always
* fail

Here is what I would imagine the signature could look like. The callback map with success and error functions is not required. 

```xquery
proc:fork($work as function(*), $arguments as item()*, callbacks as map(*))
```
On last utility method to add would then be <code>when</code>

```xquery
let $success:= function ($arguments as array?) as item()* { 
    (file:write($arguments?1,
     $arguments?2?*)
}
let $when := proc:when(($extractDoc, $extractUsers), map { 'done': $success })
return
    $when()
```

Lastly, I would imagine the forking process would be managed by a thread pool with a configurable max/min number of threads. 

These are just some ideas. Hopefully its helpful for brainstorming this idea.

**Comment by ChristianGruen** [comment]
> Hopefully its helpful for brainstorming this idea.

It is! Thanks for your valuable feedback.

**Comment by james-jw** [comment]
It works!! Take a look. The API is very much what I imagined and the pattern works beautifully. It leverages the Java 7 ForkJoinPool library and promise pattern described earlier.

Internally, it leverages the QueryModule extension point and XQFunction interface to implement the code. So far in testing it works amazingly, against remote data, local files, databases etc!

Take a look at the extensive readme for more information and let me know what you think!

### Introducing [xq-promise][0]
A library for parallelizing XQuery expressions using fork-join and promises!

[0]: https://github.com/james-jw/xq-promise

**Comment by apb2006** [comment]
Works for me also! I think this is great work.Thanks

**Comment by ChristianGruen** [comment]
James, that’s excellent work! Thanks!

I think I’ll soon check how to tackle potential dependencies and side-effects (locking of databases, updating queries, etc.).

**Comment by james-jw** [comment]
Apb and Christian, thanks for checking it out and glad its working for you too!

I am pretty amazed I was able to get this working so quickly definitely a testament to your systems design. 

One thing other thing of note. At the call to the XqPromise module methods I loose the InputInfo since the QueryModule does not appear to provide it. Thus forks are not provided it., not sure it matters aside from making it harder to debug forks.

Lastly, @apb2006 if you checked out the project yesterday, I improved the fork logic last night. In the first release half the forks just managed child forks and did not work themselves.
 
- James

**Comment by james-jw** [comment]
Made one other small tweak to the fork-join piece. Promises are not strictly required anymore as zero arity functions are accepted too.

It appears it works as well, obviously more testing is needed and this is a super simple scenario:
```xquery
  let $work := for $i in (1 to 10)
              return
                function() { trace(<value>{$i}</value>) }
  return
    let $result := promise:fork-join($work)
    return
      unit:assert-equals(sum($result), 55)
```

**Comment by ChristianGruen** [comment]
@apb2006: I’m ashamed to see I have promised to check out your solution as well in August, and in my vacation I seem to successfully have blocked it out… Work in progress ;)

**Comment by apb2006** [comment]
No problem. It is still not complete :)
 My use case is really kicking off a query and not having to wait for the
result. And  to have some way to subsequently find out how the execution
went. I need to do some work at passing data or error info on to the
subsequent executions. And actually documenting the syntax for this.

/Andy

**Comment by james-jw** [comment]
You could augment the pattern I used and add single method: ``fork`` in addition to the ``fork-join`` that is currently implemented.

The difference would be, fork would immediately initiate the process on a new thread and return a promise, which at any time in the future when asked for its value will return the result or block the thread until the result was received. This is slightly different than the queuing approach I showed, as it allows you to manually initiate forks and let the main thread continue while doing the main processing.  Finally, at the end of the main processing the results could be checked using the standard pattern if required.

```xquery
let $promise := $promise:fork($someOtherWork) 
let $mainResult :=  (: do processing on main thread :)
return
 ($mainResult, $promise(()))
```

Another difference is the call to the $promise(()) would not strictly required since the forking occurred at the call to ``$fork``. It would only be necessary if you wanted the result.

Not sure if that covers your case.

**Comment by james-jw** [comment]
Fyi, added the `fork` method as shown above. For example with the xquery:eval function you could do something like this:

```xquery
let $query := 'some xquery expression'
let $promise := promise:fork(xquery:eval(?), $query)
let $work := (: do stuff :)
return
   ($work, (: if you want return the result :) $promise(()))
```

Give it a whirl!

**Comment by apb2006** [comment]
I did, thanks. See https://github.com/james-jw/xq-promise/issues/5

**Comment by ChristianGruen** [comment]
Some more discussion on aynchronous FLWOR processing: https://github.com/james-jw/xq-promise/issues/15

**Comment by ChristianGruen** [comment]
The Async Module is now available: #1272.
